### DISCUSSION TITLE
Clock config generator

### DISCUSSION THREAD

**Discussion by twasilczyk** [modm-io/modm#715] [Ideas]
Before I dived deep into SAMV clock details, I was really afraid of *understanding* the clock tab in ASF4 (start.atmel.com) clock configurator. I used to click things around until there was no exclamation marks/warnings and the CPU clock looked high enough.

It would be useful for less-than-expert users to have a tool that would convert ASF4 configuration files (I'm sure there's an equivalent for STM) to SystemClock structure. Maybe there should be a generic SystemClock.hpp.in file for each MCU family that would get pre-filled with constants automatically extracted from `project.atstart` file by module.lb? As a bonus, the config validation would be already done by the chip maker's configuration tool.

**Comment by salkinium** [comment]
Yes, that sounds like a great idea!

I'd make that a separate [modm tool](https://github.com/modm-io/modm/tree/develop/tools/modm_tools), see [the tool documentation here](https://modm.io/reference/module/modm-build/#generic-python-tools). We can then integrate that into the build systems if needed.
